llama_stack/providers/adapters/inference/tgi/tgi.py

```diff
@@ -75,7 +78,88 @@ async def completion(
         stream: Optional[bool] = False,
         logprobs: Optional[LogProbConfig] = None,
     ) -> AsyncGenerator:
-        raise NotImplementedError()
+        request = CompletionRequest(
+            model=model,
+            content=content,
+            sampling_params=sampling_params,
+            stream=stream,
+            logprobs=logprobs,
+        )
+        if stream:
+            return self._stream_completion(request)
+        else:
+            return await self._nonstream_completion(request)
+
+    def _get_params_for_completion(self, request: CompletionRequest) -> dict:
+        prompt, input_tokens = completion_request_to_prompt_model_input_info(
+            request, self.formatter
+        )
+        max_new_tokens = min(
+            request.sampling_params.max_tokens or (self.max_tokens - input_tokens),
+            self.max_tokens - input_tokens - 1,
+        )
+
+        options = get_sampling_options(request)
+        # delete key "max_tokens" from options since its not supported by the API
+        options.pop("max_tokens", None)
+
+        if fmt := request.response_format:
+            if fmt.type == ResponseFormatType.json_schema.value:
```

Comment:
I am not sure if structured output makes sense in completions API.  This link from open AI https://cookbook.openai.com/examples/structured_outputs_intro says that its supposed to be used for chat completions and assistants api.

Reply:
I think OpenAI has generally disabled completions but we want to keep it because it is a useful primitive. Structured output most definitely still does make sense for completions() also -- it is an output modality. You are either prompting the model "raw" via completions() or via "chat" when doing chat_completion()

Reply:
got it. added a test!